course5/lec5_en.md

```diff
@@ -0,0 +1,347 @@
+  - `[3, 5, 4, 1, 2, 0]`
+- Breadth-first traversal: `[0, 1, 2, 3, 4, 5]`
+  
+![height:6em](../pics/traversal.drawio.png)
+
+# Depth-First Search: Using Search as an Example
+
+- To find a specific value in a tree's nodes
+- Structural recursion
+  - First, handle the base case: an empty tree
+  - Then, handle the recursive case and recurse
+  
+```moonbit
+fn dfs_search(target: Int, tree: IntTree) -> Bool {
+  match tree { // check the visited tree 
+    Empty => false // empty tree implies we are getting deepest
+    Node(value, left, right) => // otherwise, search in subtrees
+      value == target || dfs_search(target, left) || dfs_search(target, right)
+  }
+}
+```
+- The difference between preorder, inorder, and postorder searches is the order of operations on the current node
+
+# Short-Circuit Logical Operations
+
+- Short-circuit operation: If the result of the current evaluation can be determined, the calculation will be terminated, and the result will be returned directly
+  - `let x = true || { abort("Program stops") }`: Because `true || any value` is always true, the expression on the right side of `||` will not be evaluated
+  - `let y = false && { abort("Program stops") }`: Because `false && any value` is always false, the expression on the right side of `&&` will not be evaluated
+- Tree traversal
+  - `value == target || dfs_search(target, left) || dfs_search(target, right)` will stop traversal as soon as the target is found
+
+# Breadth-First Search
+
+- Visit each subtree on the same level one by one and in order, then proceed to visit their subtrees
+- The implementation of the algorithm relies on a First-In-First-Out data structure: queue
+  - For the trees currently in the queue, we dequeue a tree, operate on its root node, then add all of its non-empty subtrees from left to right to the end of the queue, until the queue is empty
+  
+![](../pics/bfs.drawio.png)
+
+# Data Structure: Queues
+
+- Similar to queuing in everyday life, the first person to enter the queue gets served first
+- The insertion and removal of data follow the First-In-First-Out (FIFO) principle
+  - Data is inserted at the back of the queue and removed from the front
+  
+![](../pics/queue.drawio.png)
+
+# Data Structure: Queues
+
+- The queue we use here is defined by the following interfaces:
+  
+```moonbit no-check
+fn empty[T]() -> Queue[T] // construct an empty queue
+fn enqueue[T](q: Queue[T], x: T) -> Queue[T] // add element to the tail
+// attempt to dequeue an element, return None if the queue is empty
+fn pop[T](q: Queue[T]) -> (Option[T], Queue[T]) 
+```
+
+- For example
+  
+```moonbit no-check
+let q = enqueue(enqueue(empty(), 1), 2)
+let (head, tail) = pop(q)
+assert(head == Some(1))
+assert(tail == enqueue(empty(), 2))
+```
+
+# Breadth-First Search: Using Search as an Example
+
+- To find a specific value in a tree's nodes
+  
+```moonbit no-check
```

Comment:
why are we having `no-check` here?

Reply:
Because the definition of Queue is abstract. Do I need to complete the implementation of `Queue` and its operations at #167?

Reply:
Well, what you can do is
```markdown
<!-- create a comment area
--- moonbit
// and put the implementation inside
---
-->
```

Reply:
Please check [bef1eed](https://github.com/moonbitlang/moonbit-course/pull/15/commits/bef1eed879c93a5c1d186acb4392fe9804477626).